src/eos-updater-avahi.c

```diff
@@ -0,0 +1,678 @@
+  discoverer->callback = callback;
+  discoverer->user_data = user_data;
+  discoverer->notify = notify;
+  if (context == NULL)
+    discoverer->context = g_main_context_ref_thread_default ();
+  else
+    discoverer->context = g_main_context_ref (context);
+  discoverer->discovered_services = g_hash_table_new_full (g_str_hash,
+                                                      g_str_equal,
+                                                      g_free,
+                                                      NULL);
+  discoverer->found_services = object_array_new ();
+
+  discoverer->poll = avahi_glib_poll_new (context, G_PRIORITY_DEFAULT);
+  discoverer->client = avahi_client_new (avahi_glib_poll_get (discoverer->poll),
+                                         0,
+                                         client_cb,
+                                         discoverer,
+                                         &failure);
+
+  if (discoverer->client == NULL)
+    {
+      g_set_error (error,
+                   EOS_UPDATER_ERROR,
+                   EOS_UPDATER_ERROR_LAN_DISCOVERY_ERROR,
+                   "Failed to create discoverer client: %s",
+                   avahi_strerror (failure));
+      return FALSE;
+    }
+
+  discoverer->browser = avahi_service_browser_new (discoverer->client,
+                                                   AVAHI_IF_UNSPEC,
+                                                   AVAHI_PROTO_UNSPEC,
+                                                   EOS_UPDATER_AVAHI_SERVICE_TYPE,
+                                                   NULL,
+                                                   0,
+                                                   browse_cb,
+                                                   discoverer);
+  if (discoverer->browser == NULL)
+    {
+      g_set_error (error,
+                   EOS_UPDATER_ERROR,
+                   EOS_UPDATER_ERROR_LAN_DISCOVERY_ERROR,
+                   "Failed to create service browser: %s",
+                   avahi_strerror (avahi_client_errno (discoverer->client)));
+      return NULL;
+    }
+  return g_steal_pointer (&discoverer);
+}
+
+static GBytes *
+generate_from_template (const gchar *tmpl,
+                        GHashTable *values,
+                        GError **error)
+{
+  g_auto(GStrv) splitted = g_strsplit (tmpl, "@", -1);
+  gchar **iter;
+  gboolean special = FALSE;
+  g_autoptr(GString) xml = g_string_new (NULL);
+
+  for (iter = splitted; *iter != NULL; ++iter)
+    {
+      if (special)
+        {
+          const gchar *value = g_hash_table_lookup (values, *iter);
+
+          if (value == NULL)
+            {
+              g_set_error (error, G_IO_ERROR, G_IO_ERROR_FAILED,
+                           "No value provided for token %s", *iter);
+              return NULL;
+            }
+          g_string_append (xml, value);
+        }
+      else
+        g_string_append (xml, *iter);
```

Comment:
This looks like a pretty efficient way of doing find-and-replace on tokens in a template string, though the implementation took me a minute or two to parse. 

I wonder if it would be worth adding a little comment above which says that we're splitting "template-like" strings (e.g., those with with @FOO@ into their components and that every even component after an @ is guaranteed to be a template-key that needs to be replaced.